src/components/tabs/ScatterPlotTab.tsx

```diff
@@ -386,6 +392,131 @@ export default memo(function ScatterPlotTab(props: ScatterPlotTabProps): ReactEl
+
+    const traceDataBuckets: TraceData[] = [];
+    if (!featureData || markerConfig.color || overrideColor) {
+      // Do no coloring! Keep all points in the same bucket.
+      traceDataBuckets.push({
+        x: Array.from(xData),
+        y: Array.from(yData),
+        objectIds: objectIds,
+        trackIds: trackIds,
+        color: overrideColor || new Color(),
+      });
+    } else {
+      // Generate colors
+      const categories = dataset.getFeatureCategories(selectedFeatureName);
+      const isCategorical = categories !== null;
+      const colors = isCategorical
+        ? categoricalPalette.slice(0, categories.length)
+        : subsampleColorRamp(colorRamp, COLOR_RAMP_SUBSAMPLES);
+      const colorMinValue = isCategorical ? 0 : colorRampMin;
+      const colorMaxValue = isCategorical ? categories.length - 1 : colorRampMax;
+
+      // Make a bucket group for each ramp/palette color and for the out-of-range and outliers.
+      // index 0 = out of filter range, index 1 = outliers.
+      for (let i = 0; i < colors.length + 2; i++) {
+        let color;
+        if (i === 0) {
+          color = viewerConfig.outOfRangeDrawSettings.color;
+        } else if (i === 1) {
+          color = viewerConfig.outlierDrawSettings.color;
+        } else {
+          color = colors[i - 2];
+        }
+        traceDataBuckets.push({ x: [], y: [], objectIds: [], trackIds: [], color: color });
+      }
+
+      // Sort data into buckets
+      for (let i = 0; i < xData.length; i++) {
+        const isOutOfRange = false; // TODO: Implement out of range filtering
+        const isOutlier = dataset.outliers ? dataset.outliers[objectIds[i]] : false;
+        const objectId = objectIds[i];
+
+        let bucketIndex;
+        if (isOutOfRange) {
+          bucketIndex = 0;
+        } else if (isOutlier) {
+          bucketIndex = 1;
+        } else {
+          bucketIndex = getBucketIndex(featureData.data[objectId], colorMinValue, colorMaxValue, colors.length) + 2;
+        }
+
+        const bucket = traceDataBuckets[bucketIndex];
+        bucket.x.push(xData[i]);
+        bucket.y.push(yData[i]);
+        bucket.objectIds.push(objectIds[i]);
+        bucket.trackIds.push(trackIds[i]);
+      }
+
+      // Optionally delete the outlier and out of range buckets to hide the values.
+      if (viewerConfig.outlierDrawSettings.mode === DrawMode.HIDE) {
+        traceDataBuckets.splice(1, 1);
+      }
+      if (viewerConfig.outOfRangeDrawSettings.mode === DrawMode.HIDE) {
+        traceDataBuckets.splice(0, 1);
+      }
+    }
+
+    // Transform buckets into traces
+    const traces: Partial<PlotData>[] = traceDataBuckets
+      .filter((bucket) => bucket.x.length > 0) // Remove empty buckets
+      .reduce((acc: TraceData[], bucket: TraceData) => {
+        // Split the traces into smaller chunks to prevent plotly from freezing.
+        acc.push(...splitTraceData(bucket, MAX_POINTS_PER_TRACE));
+        return acc;
+      }, [])
```

Comment:
would love it if I could recommend `flatMap` here... would look so much nicer... would (I think) be less efficient...

Reply:
This made me cackle out loud, thank you